addon/-private/collapse-tree.js

```diff
@@ -0,0 +1,385 @@
+/* global Ember */
+import { get, set } from '@ember/object';
+import { assert } from '@ember/debug';
+
+import { computed } from '@ember-decorators/object';
+import { addObserver } from '@ember/object/observers';
+
+/**
+  Genericizes `objectAt` so it can be run against a normal array or an Ember array
+
+  @param {object|Array} arr
+  @param {number} index
+  @return {any}
+*/
+function objectAt(arr, index) {
+  assert('arr must be an instance of a Javascript Array or implement `objectAt`', Array.isArray(arr) || typeof arr.objectAt === 'function');
+
+  if (typeof arr.objectAt === 'function') {
+    return arr.objectAt(index);
+  }
+
+  return arr[index];
+}
+
+/**
+  Given a list of ordered values and a target value, finds the index of
+  the closest value which does not exceed the target value
+
+  @param {Array<number>} values - the list of values
+  @param {number} target - the index to find the closest value to
+  @return {number} - the index of the value closest to the target
+*/
+function closestLessThan(values, target) {
+  let low = 0;
+  let high = values.length - 1;
+
+  while (low <= high) {
+    let mid = Math.floor((high + low) / 2);
+
+    if (target < values[mid]) {
+      high = mid - 1;
+    } else if (target > values[mid]) {
+      low = mid + 1;
+    } else {
+      return mid;
+    }
+  }
+
+  // low === high + 1, we always want the lower value
+  return high;
+}
+
+/**
+  Single node of a CollapseTree
+*/
+class Node {
+  constructor(value, parent) {
+    assert('value must have an array of children', Array.isArray(get(value, 'children')));
+
+    set(this, 'value', value);
+
+    if (parent) {
+      // Changes to the value directly should properly update all computeds on this
+      // node, but we need to manually propogate changes upwards to notify any other
+      // watchers
+      addObserver(this, 'length', () => Ember.propertyDidChange(parent, 'length'));
```

Comment:
Does this create memleak when Node is no longer used?

Reply:
Observers are instance state and will be garbage collected along with the instance, so long as there are no other references to the instance.